### PR TITLE
feat(Commands): resolve mentions into objects

### LINKED ISSUE #190: [Feature Request] Argument parser improvements
**What do you want to request?**
A few improvements with the arg parser:
 - Parses users into User objects, roles into Role objects etc
 - Also handles names, could return `false` if theres roles/users/channels with the same name so the bot dev can accurately send out an error message.

### PR DESCRIPTION
**Pull request by MierenManz**
## About

this should close #190
Feedback is appreciated
cc @DjDeveloperr and @grian32

## Status

- [x] These changes have been tested against Discord API or do not contain API change.
- [ ] This PR includes only documentation changes, no code change.
- [x] This PR introduces some Breaking changes.

### CODE REVIEW

**Comment by grian32** [comment]
Does this also resolve ID's into objects?

**Comment by MierenManz** [comment]
> Does this also resolve ID's into objects?
It should do that
```ts
const mentionToRegex: MentionToRegex = {
  user: /<@!?(\d{17,19})>|(\d{17,19})/,
  role: /<@&(\d{17,19})>|(\d{17,19})/,
  channel: /<#(\d{17,19})>|(\d{17,19})/
}
```
These are the regex's

**Comment by MierenManz** [comment]
I hate lint!

**Comment by MierenManz** [comment]
@Helloyunho @grian32 and @DjDeveloperr this is now ready for review
I appreciate all feedback on the test cases. Especially because the mocking can be kinda weird.

I added a few comments to everything and I hope that explain enough.
But please if you have any questions. Just put them in the review :D

**Comment by MierenManz** [comment]
> Does this also resolve ID's into objects?

There was a flaw in the code where it wouldn't solve that properly. Now it does tho :D

**Review by grian32 — APPROVED** [review]
LGTM, I've also been able to test locally with my bot and it works as expected, good job!

**Comment by MierenManz** [comment]
> LGTM, I've also been able to test locally with my bot and it works as expected, good job!

thanks :3 Wasn't sure if the mocking was accurate.
Seems like it is and I hope that it is good enough

File: src/utils/command.ts [hunk 1]
```diff
@@ -60,7 +69,13 @@ export function parseArgs(
       case 'user':
       case 'role':
       case 'channel':
-        parseMention(args, entry, messageArgsNullableCopy)
+        await parseMention(
+          args,
+          entry,
+          messageArgsNullableCopy,
+          message.client,
```
Comment (Helloyunho) [inline src/utils/command.ts:76]: maybe use one message arg right below and get client in `parseMention`

Comment (MierenManz) [reply src/utils/command.ts:76]: Like don't pass message.client as parameter.
But replace `client` in the code with `message.client`?

Comment (Helloyunho) [reply src/utils/command.ts:76]: yes

**Review by Helloyunho — APPROVED** [review]
lgtm! let's wait for dj

File: src/utils/command.ts [hunk 2]
```diff
@@ -43,12 +48,15 @@ export type Args =
...
-  messageArgs: string[]
-): Record<string, string | number | boolean> | null {
+  message: Message
+): Promise<Record<
+  string,
+  Guild | User | Role | string | number | boolean
+> | null> {
   if (commandArgs === undefined) return null
-
+  const messageArgs = message.content.split(' ')
```
Comment (DjDeveloperr) [inline src/utils/command.ts:59]: I'm confused what is this for. Don't we already parse arguments into a raw argument part string (`argString`) and arguments array (`rawArgs`)?

Comment (MierenManz) [reply src/utils/command.ts]: fixed it.

**Review by DjDeveloperr — APPROVED** [review]
LGTM!

**Pull request merged**